projects: add ORAS proposal (#68)
opencontainers/tob: proposals/oras.md

# OCI ORAS Project Proposal #

## Abstract ##

The ORAS project (https://github.com/deislabs/oras) is a CLI that can publish arbitrary content to an OCI registry, with special features for setting mediatypes on manifest configs and on content. 

In order to provide OCI end users a method to publish and retrieve any type of content to/from an OCI registry, as well as a reference implementation for the in-progress artifacts spec, ORAS should be moved under the opencontainers GitHub org.

### ORAS Details ###

> 

Reviewer: Markdown only requires 1 line after titles not 2
```suggestion
```

Example - uploading rockets, a brand new type of package:

```
# Create a thing
printf '🚀' > rocket.txt

# Create a manifest config
printf '{"RocketVersion":"v0.1.0"}' > rocket-config.json

# Upload your thing with a custom mediatype
oras push localhost:5000/mystuff/myrocket:v0.1.0 rocket.txt:text/plain \
  --manifest-config rocket-config.json:application/vnd.acme.rocket.config.v1+json
 ```

See manifest created:

```
$ curl -s -H 'Accept: application/vnd.oci.image.manifest.v1+json' \
    http://localhost:5000/v2/mystuff/myrocket/manifests/v0.1.0 | jq
{
  "schemaVersion": 2,
  "config": {
    "mediaType": "application/vnd.acme.rocket.config.v1+json",
    "digest": "sha256:310175f34d2d4d5cba3418be06ddd1ef948147d729516d78318ec7f5c2d83d49",
    "size": 26
  },
  "layers": [
    {
      "mediaType": "text/plain",
      "digest": "sha256:ebbc0b2870eb323f2b6cffa5c493ceef81ae7eb36afc73d4e0367301631daec5",
      "size": 4,
      "annotations": {
        "org.opencontainers.image.title": "rocket.txt"
      }
    }
  ]
}
```

Get that thing:

```
$ curl -s http://localhost:5000/v2/mystuff/myrocket/blobs/sha256:ebbc0b2870eb323f2b6cffa5c493ceef81ae7eb36afc73d4e0367301631daec5
🚀
```

> 

Reviewer: ```suggestion
```

	rocketConfigMediaType := "application/vnd.acme.rocket.config.v1+json"
	rocketConfigContent := []byte(fmt.Sprintf("{\"RocketVersion\":\"%s\"}", rocketVersion))
	rocketConfigDescriptor := store.Add("", rocketConfigMediaType, rocketConfigContent)

	ref := fmt.Sprintf("%s/%s:%s", registryRootURL, registryNamespace, rocketVersion)
	_, err := oras.Push(ctx, resolver, ref, store, []ocispec.Descriptor{rocketDescriptor},
		oras.WithConfig(rocketConfigDescriptor))
	if err != nil {
		panic(err)
	}

	fmt.Println("Pushed to", ref)
	fmt.Printf("\nTry:\n\ncurl -s -H 'Accept: application/vnd.oci.image.manifest.v1+json' \\\n" +
		"    %s/v2/%s/manifests/%s | jq\n", registryRootURL, registryNamespace, rocketVersion)
}
```

You can see all features in the project [README](https://github.com/deislabs/oras/blob/master/README.md).

#### Adoption ####

The following projects are already successfully using ORAS to work with custom artifacts:

- [Helm](https://github.com/helm/helm/search?q=oras)
- [Conftest](https://github.com/instrumenta/conftest/search?q=oras)
- [Singularity](https://github.com/sylabs/singularity/search?q=oras)

> 

Reviewer: ```suggestion
```

## Proposal ##

> Change the ownership of the existing umoci project from deislabs:

Reviewer 2: `s/umoci/oras/g`. :wink:

Author: :)

  https://github.com/deislabs/oras

And move it inside the `opencontainers` organization:

  https://github.com/opencontainers/oras

The import paths will correspondingly be "github.com/opencontainers/oras" (oras does have some Go API users, but since the project will be renamed -- and GitHub will add a redirect -- there will be no significant downstream impact of the change).

### Initial Maintainers ###
Initial maintainers of the ORAS project would be:

* Josh Dolitsky <josh@bloodorange.io> (@jdolitsky)
* Shiwei Zhang <shizh@microsoft.com> (@shizhMSFT)
* Sajay Antony <sajaya@microsoft.com> (@sajayantony)
* Steve Lasker <steve.lasker@microsoft.com> (@stevelasker)
* Jimmy Zelinskie <jimmyzelinskie@gmail.com> (@jzelinskie)
* Matt Fisher <matt.fisher@microsoft.com> (@bacongobbler)

### Code of Conduct ###
This project would incorporate (by reference) the OCI [Code of Conduct][code-of-conduct].

[code-of-conduct]: https://github.com/opencontainers/org/blob/master/CODE_OF_CONDUCT.md

### Governance and Releases ###

> This project would incorporate the Governance and Releases processes from the OCI project template: https://github.com/opencontainers/project-template.

Reviewer 2: Should be a markdown link, not a raw URL.

It should be noted that since ORAS is not a specification, it is not bound by the ordinary quorum and voting rules for specification release.
As such, new versions will be released as regularly as needed without the need for a quorum vote.

> Until there are enough additional maintainers, PRs will be merged with only one maintainer LGTM required.

Reviewer 2: This doesn't quite appear to match the maintainers list (which has a fair number of people) -- two LGTMs shouldn't be an unreasonable requirement for PRs.

Reviewer: 2  ++

Maintainers are encouraged to not merge their own PRs immediately (preferably waiting at least 24 to 48 hours to allow another maintainer to review it), but this is not a strict rule.

### Project Communications ###
The proposed project would continue to use existing channels in use by the OCI developer community for communication including:

 * GitHub for issues and pull requests.
 * The [`dev@opencontainers.org`][oci-ml] email list.
 * The weekly OCI developer community conference call.
 * The `#opencontainers` IRC channel on Freenode.
 * The [OCI Slack workspace][oci-slack].
 * The [OCI Matrix Room][oci-matrix].

[oci-ml]: mailto:dev@opencontainers.org
[oci-slack]: https://opencontainers.slack.com/
[oci-matrix]: https://matrix.to/#/#opencontainers:matrix.org

> 

Reviewer 2: FAQ section is missing.

# OCI ORAS Project Proposal #

## Abstract ##

The [ORAS project][oras-project] is a CLI that can publish arbitrary content to an [OCI Spec Compliant](https://github.com/opencontainers/oci-conformance) registry, with special features for setting mediatypes on manifest configs and on content. 

In order to provide OCI end users a method to publish and retrieve any type of content to/from an OCI registry, as well as a reference implementation for the artifacts project, ORAS should be moved under the opencontainers GitHub org.

[oras-project]: https://github.com/deislabs/oras

### ORAS Details ###

> ORAS is a CLI that can publish arbitrary content to an OCI registry, with special features for setting mediatypes on manifest configs and on content.

Author: yes, you're right. I have gone ahead and kept it here and simplified the first sentence

Note: the manifest mediatype itself is always `application/vnd.oci.image.manifest.v1+json`.

Example - uploading rockets, a brand new type of package:

```
# Create a thing
printf '🚀' > rocket.txt

# Create a manifest config
printf '{"RocketVersion":"v0.1.0"}' > rocket-config.json

# Upload your thing with a custom mediatype
oras push localhost:5000/mystuff/myrocket:v0.1.0 rocket.txt:text/plain \
  --manifest-config rocket-config.json:application/vnd.acme.rocket.config.v1+json
 ```

See manifest created:

```
$ curl -s -H 'Accept: application/vnd.oci.image.manifest.v1+json' \
    http://localhost:5000/v2/mystuff/myrocket/manifests/v0.1.0 | jq
{
  "schemaVersion": 2,
  "config": {
    "mediaType": "application/vnd.acme.rocket.config.v1+json",
    "digest": "sha256:310175f34d2d4d5cba3418be06ddd1ef948147d729516d78318ec7f5c2d83d49",
    "size": 26
  },
  "layers": [
    {
      "mediaType": "text/plain",
      "digest": "sha256:ebbc0b2870eb323f2b6cffa5c493ceef81ae7eb36afc73d4e0367301631daec5",
      "size": 4,
      "annotations": {

>         "org.opencontainers.image.title": "rocket.txt"

Author: This is the default. It is basically the way oras knows how to write files to the current directory on pull.

However, this can be disabled by using custom layers/descriptors and `oras.WithNameValidation(nil)` on push.

The annotation itself was take from image-spec, see [here](https://github.com/opencontainers/image-spec/blob/da296dcb1e473a9b4e2d148941d7faa9ac8fea3f/annotations.md).

I wouldn't consider this a part of any spec surrounding artifacts, its just the way oras CLI chooses to store filenames. In the case of Helm, for example, it is entirely unused despite using oras as a library.

>         "org.opencontainers.image.title": "rocket.txt"

Reviewer 3: The annotations are something we've debated. I'm not a big fan of them, and these should likely be something that went into the config for a specific artifact type. That said, it's interesting to see what's possible.

      }
    }
  ]
}
```

Get that thing:

```
$ curl -s http://localhost:5000/v2/mystuff/myrocket/blobs/sha256:ebbc0b2870eb323f2b6cffa5c493ceef81ae7eb36afc73d4e0367301631daec5
🚀
```

#### Additional Usage ####

ORAS is built primarily on top of Go packages provided by [containerd][containerd-project], but it also imports packages from the [docker/cli][dockercli-project], which enables "docker-style" auth login:

```
oras login -u username -p password localhost:5000 -c rocket-creds.json
```

There are also public Go packages available to build on top of ORAS. The following is the equivalent of the rocket example with the CLI above, but in Go:

```go
package main

import (
	"context"
	"fmt"

	"github.com/containerd/containerd/remotes/docker"
	"github.com/deislabs/oras/pkg/content"
	"github.com/deislabs/oras/pkg/oras"
	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
)

func main() {
	ctx := context.Background()

> 	resolver := docker.NewResolver(docker.ResolverOptions{})

Author: I've just opened https://github.com/deislabs/oras/issues/165 to track this

It should be noted that since ORAS is not a specification, it is not bound by the ordinary quorum and voting rules for specification release.
As such, new versions will be released as regularly as needed without the need for a quorum vote.

Pull requests will require two (2) reviews (signified by "LGTM") from project maintainers.
Maintainers are not allowed to review a pull request which they authored.

[oci-project-template]: https://github.com/opencontainers/project-template

### Project Communications ###
The proposed project would continue to use existing channels in use by the OCI developer community for communication including:

 * GitHub for issues and pull requests.
 * The [`dev@opencontainers.org`][oci-ml] email list.
 * The weekly OCI developer community conference call.
 * The `#opencontainers` IRC channel on Freenode.
 * The [OCI Slack workspace][oci-slack].
 * The [OCI Matrix Room][oci-matrix].

[oci-ml]: mailto:dev@opencontainers.org
[oci-slack]: https://opencontainers.slack.com/
[oci-matrix]: https://matrix.to/#/#opencontainers:matrix.org

## Frequently Asked Questions (FAQ)
*Q: Does this change the OCI Charter or Scope Table?*

> A: Yes, this will add an additional row to the [OCI Scope Table][oci-scope-table], which does not currently include any mention of generic OCI image tooling.

Reviewer 2: This line needs to be updated now that there isn't an OCI Scope Table anymore (as I did in #67). In addition, ORAS should probably be referred to as an "OCI distribution tool"?

You should also answer some of the new questions I've added to #67 -- which effectively boils down to "what category of project is this (as defined by #76 and the still-draft #86)?". The reason this is important is that it's not quite clear whether this intends to be a Reference Implementation of *something* (whether that be distribution-spec as a client, or artifacts) or a Library, and the scope and usage of the project depends a fair bit on that categorisation. My view is that this appears to be a Reference Implementation (because while it is usable as a Library that's effectively an implementation detail -- umoci is also usable as a library, as is runc).

You also should explain how you plan to avoid the "runc issue" (conventions in a Reference Implementation that are within the scope of the Specification becoming a de-facto standard even though they weren't standardised and thus are impossible to change).

Author: Just updated FAQ section, stating ORAS a reference implementation of distribution spec

Reviewer 3: We've also said it was a library, to be used by other tools. I suppose the question is whether it can be in both categories.

Reviewer 3: @cyphar It looks like Josh's update addresses the feedback. Is there something else to clear the requested change?

}
```

You can see all features in the project [README][oras-readme].

[containerd-project]: https://github.com/containerd/containerd
[dockercli-project]: https://github.com/docker/cli
[oras-readme]: https://github.com/deislabs/oras/blob/master/README.md

#### Adoption ####

The following projects are already successfully using ORAS to work with custom artifacts:

- [Helm][helm-usage]
- [Conftest][conftest-usage]
- [Singularity][singularity-usage]

[helm-usage]: https://github.com/helm/helm/search?q=oras
[conftest-usage]: https://github.com/instrumenta/conftest/search?q=oras
[singularity-usage]: https://github.com/sylabs/singularity/search?q=oras

## Proposal ##
Change the ownership of the existing ORAS project from deislabs:

  https://github.com/deislabs/oras

And move it inside the `opencontainers` organization:

  https://github.com/opencontainers/oras

The import paths will correspondingly be "github.com/opencontainers/oras" (oras does have some Go API users, but since the project will be renamed -- and GitHub will add a redirect -- there will be no significant downstream impact of the change).

### Initial Maintainers ###

> Initial maintainers of the ORAS project would be:

Reviewer 4: A) This may be too many folks from a single employer
B) I could participate here

Reviewer 3: Since you're no longer at RedHat, that would make 4 companies.

Reviewer 3: We did reduce Matt Fisher - see pr https://github.com/deislabs/oras/pull/162

* Josh Dolitsky <josh@bloodorange.io> (@jdolitsky)
* Shiwei Zhang <shizh@microsoft.com> (@shizhMSFT)
* Sajay Antony <sajaya@microsoft.com> (@sajayantony)
* Steve Lasker <steve.lasker@microsoft.com> (@stevelasker)
* Jimmy Zelinskie <jimmyzelinskie@gmail.com> (@jzelinskie)
* Matt Fisher <matt.fisher@microsoft.com> (@bacongobbler)

### Code of Conduct ###
This project would incorporate (by reference) the OCI [Code of Conduct][code-of-conduct].

[code-of-conduct]: https://github.com/opencontainers/org/blob/master/CODE_OF_CONDUCT.md

### Governance and Releases ###
This project would incorporate the Governance and Releases processes from the [OCI project template][oci-project-template].

It should be noted that since ORAS is not a specification, it is not bound by the ordinary quorum and voting rules for specification release.
As such, new versions will be released as regularly as needed without the need for a quorum vote.

> PRs will require two (2) LGTMs from project maintainers

Reviewer 2: Two points:

 1. You should clarify whether maintainers are allowed to "self-LGTM" (as in, are they allowed to LGTM a pull request which they authored)? Technically this is not permitted by the OCI project template, but it's up to each OCI project to decide whether this is a reasonable maintainership model.

 2. This should probably be written as:

```suggestion
Pull requests will require two (2) reviews (signified by "LGTM") from project maintainers.
```

> Maintainers are encouraged to not merge their own PRs immediately (preferably waiting at least 24 to 48 hours to allow another maintainer to review it), but this is not a strict rule.

Reviewer 2: This line is no longer necessary -- by definition you can't merge your own PR without other people reviewing it if you require two LGTMs.

Pull requests will require two (2) reviews (signified by "LGTM") from project maintainers.
Maintainers are not allowed to review a pull request which they authored.

[oci-project-template]: https://github.com/opencontainers/project-template

### Project Communications ###
The proposed project would continue to use existing channels in use by the OCI developer community for communication including:

 * GitHub for issues and pull requests.
 * The [`dev@opencontainers.org`][oci-ml] email list.
 * The weekly OCI developer community conference call.
 * The `#opencontainers` IRC channel on Freenode.
 * The [OCI Slack workspace][oci-slack].
 * The [OCI Matrix Room][oci-matrix].

[oci-ml]: mailto:dev@opencontainers.org
[oci-slack]: https://opencontainers.slack.com/
[oci-matrix]: https://matrix.to/#/#opencontainers:matrix.org

## Frequently Asked Questions (FAQ)
> *Does this proposal change the OCI Charter?*

This proposal does not in any way intend to amend the [OCI Charter][oci-charter].

[oci-charter]: https://github.com/opencontainers/tob/blob/master/CHARTER.md

> *Where does ORAS fit into the OCI suite of projects?*

> ORAS is intended to be a *reference implementation of the OCI Distribution Specification*.

Reviewer 5: This is really hard for me to wrap my head around.  As far as I can understand, containerd's `dockerResolver` seems to be the actual OCI distribution client that is being used as part of ORAS.

It seems to me that ORAS is more of a library and tool for enabling artifact-related workflows with an OCI distribution.

Author: Yes, I sort of agree here.. I was going off of Aleksa's suggestion. 

Any issue marking this as a library vs. reference implementation? Does marking it one way or another have any effect? @cyphar @SteveLasker

Reviewer 2: Yes, there is a distinction. From my point of view, ORAS cannot be argued to be a library because it isn't small enough nor is it merely a tool to assist the implementation of OCI specifications. It is (ostensibly) a (reference) implementation of the distribution-spec that happens to have a bunch of other features. You could argue that the `dockerResolver` of containerd is also a (reference) implementation of the distribution-spec. Not to mention it has a CLI, but I don't want to give the impression that that is the sticking point.

You could argue we need another category for "things that aren't reference implementations" but I think that would dilute the mission of the OCI too far. If you both feel that ORAS could be argued as a library then you should comment on #76 (and eventually #86) to express your disagreement with the current text of both documents.

Reviewer 3: ORAS was created as a library to support additional artifacts with oci-distribution. Pushing pulling content, creating manifests, .tar-ing up contents isn't trivial for the average person wanting to focus on their unique artifact type and benefit from a registry.

Sure, it happens to implement the spec, but I don't know if I'd call it a reference implementation as mentioned, it does use other libraries as well. We didn't think it needed to duplicate what functionality already existed. 

So, happy to see the text changed to be a library, intended for artifact usage. 
I'd suggest we need libraries with focus on container runtimes, but we can also benefit from libraries that leverage the generic nature of registries to support all types of artifacts.

Reviewer 5: @cyphar Can you help me understand more about how you see ORAS as an implementation of the distribution-spec?  I can see how ORAS enables people to work more easily with distributions, but ORAS itself (at least in my understanding) doesn't implement the distribution protocol.

As ORAS was designed to handle any type of content, it will serve to exercise the spec
to make it more independent of details that may have leaked in from image-spec or Docker.

> *Why bless ORAS over other alternative tools?*

ORAS has already been used successfully in the wild as a building block for 
publishing custom content to an OCI registry.

The following projects are already successfully using ORAS to work with custom artifacts:

- [Helm][helm-usage]
- [Conftest][conftest-usage]
- [Singularity][singularity-usage]

[helm-usage]: https://github.com/helm/helm/search?q=oras
[conftest-usage]: https://github.com/instrumenta/conftest/search?q=oras
[singularity-usage]: https://github.com/sylabs/singularity/search?q=oras

> *How do we avoid the runc issue with implementation-specific quirks becoming a de-facto standard?*

When developing new features in ORAS (which are within the scope of the OCI Distribution Specification),

> a strong effort will be made to include those features in the upstream specification.

Reviewer 5: Can ORAS commit to ensuring that non-specified features remain marked as strictly experimental and unsupported until such time as those features are included in the upstream specification?

Author: @samuelkarp sure. Meaning from a readme/documentation perspective? I'm assuming this in reference to keying off of `config.mediaType` for determining artifact types

Reviewer 5: Yes, and discouraging use.  The concern I have here is the same as the concern about the annotations: I'm worried about non-standardized behavior of a reference implementation becoming a de-facto requirement.

Reviewer 2: I would go further than just documentation. You should have to pass an actual `--unsupported` or `--experimental` or `--i-understand-this-voids-my-warranty` flag. That is what I intend to do for umoci, and given my experience with documentation not being read by any user ever (in the case of `runc` at least), I think putting text in documentation is mostly a CYA thing than actually having an impact on user behaviour.

Reviewer 3: Marking experiments that aren't in a versioned/released spec as experimental makes sense. As we enable Notary v2, I expect a bunch of experiments that we'd likely enable.
That said, I don't understand the question on the `config.mediaType`. Can someone elaborate the concern here?

Reviewer 3: Talking with Sam, the question was related to the implicit use of annotations, not `config.mediatype` and how the use of annotations could imply a standard, instead of something used for the bin prototype.  And, the experimental references which we agree with as well.
We'll discuss removing the annotations, possibly placing in the config object. Tracked at https://github.com/deislabs/oras/issues/167
We'd likely leave the helper capabilities for artifact authors to add annotations for their specific usage, but avoid always generating, implying a standard.
